*** gcc-10.2.0.org/gcc/config.gcc	2021-10-16 10:31:18.000000000 +0900
--- gcc-10.2.0.z80-elf.mine/gcc/config.gcc	2022-04-06 08:00:17.000000000 +0900
***************
*** 1677,1682 ****
--- 1677,1686 ----
  	tmake_file="h8300/t-h8300"
  	tm_file="h8300/h8300.h dbxelf.h elfos.h newlib-stdint.h h8300/elf.h"
  	;;
+ z80-*-elf*)
+ 	tmake_file="z80/t-z80"
+ 	tm_file="z80/z80.h dbxelf.h elfos.h"
+ 	;;
  h8300-*-linux*)
  	tmake_file="${tmake_file} h8300/t-h8300 h8300/t-linux"
  	tm_file="h8300/h8300.h dbxelf.h elfos.h gnu-user.h linux.h glibc-stdint.h h8300/linux.h"
