*** gdb-10.2.aarch64-elf/gdb/bfin-tdep.c	2021-04-25 13:06:26.000000000 +0900
--- gdb-10.2.bfin-elf/gdb/bfin-tdep.c	2021-07-28 07:26:01.000000000 +0900
***************
*** 667,674 ****
  		   gdb_byte *readbuf,
  		   const gdb_byte *writebuf)
  {
!   if (TYPE_LENGTH (type) > 8)
!     return RETURN_VALUE_STRUCT_CONVENTION;
  
    if (readbuf)
      bfin_extract_return_value (type, regcache, readbuf);
--- 667,680 ----
  		   gdb_byte *readbuf,
  		   const gdb_byte *writebuf)
  {
!   if (TYPE_LENGTH (type) > 8) {
!     if (readbuf) {
!       ULONGEST addr;
!       regcache_raw_read_unsigned (regcache, BFIN_R0_REGNUM, &addr);
!       read_memory (addr, readbuf, TYPE_LENGTH(type));
!     }
!     return RETURN_VALUE_ABI_RETURNS_ADDRESS;
!   }
  
    if (readbuf)
      bfin_extract_return_value (type, regcache, readbuf);
