*** gdb-10.2.h8300-elf/gdb/h8300-tdep.c	2021-07-27 16:20:23.000000000 +0900
--- gdb-10.2.aarch64-elf/gdb/h8300-tdep.c	2021-04-25 13:06:26.000000000 +0900
***************
*** 880,893 ****
  		    struct type *type, struct regcache *regcache,
  		    gdb_byte *readbuf, const gdb_byte *writebuf)
  {
!   if (h8300_use_struct_convention (type)) {
!     if (readbuf) {
!       ULONGEST addr;
!       regcache_raw_read_unsigned (regcache, E_RET0_REGNUM, &addr);
!       read_memory (addr, readbuf, TYPE_LENGTH(type));
!     }
!     return RETURN_VALUE_ABI_RETURNS_ADDRESS;
!   }
    if (writebuf)
      h8300_store_return_value (type, regcache, writebuf);
    else if (readbuf)
--- 880,887 ----
  		    struct type *type, struct regcache *regcache,
  		    gdb_byte *readbuf, const gdb_byte *writebuf)
  {
!   if (h8300_use_struct_convention (type))
!     return RETURN_VALUE_STRUCT_CONVENTION;
    if (writebuf)
      h8300_store_return_value (type, regcache, writebuf);
    else if (readbuf)
