*** gdb-10.2.aarch64-elf/gdb/iq2000-tdep.c	2021-04-25 13:06:26.000000000 +0900
--- gdb-10.2.iq2000-elf/gdb/iq2000-tdep.c	2021-07-28 10:04:29.000000000 +0900
***************
*** 564,571 ****
  		     struct type *type, struct regcache *regcache,
  		     gdb_byte *readbuf, const gdb_byte *writebuf)
  {
!   if (iq2000_use_struct_convention (type))
!     return RETURN_VALUE_STRUCT_CONVENTION;
    if (writebuf)
      iq2000_store_return_value (type, regcache, writebuf);
    else if (readbuf)
--- 564,577 ----
  		     struct type *type, struct regcache *regcache,
  		     gdb_byte *readbuf, const gdb_byte *writebuf)
  {
!   if (iq2000_use_struct_convention (type)) {
!     if (readbuf) {
!       ULONGEST addr;
!       regcache_raw_read_unsigned (regcache, E_FN_RETURN_REGNUM, &addr);
!       read_memory (addr, readbuf, TYPE_LENGTH(type));
!     }
!     return RETURN_VALUE_ABI_RETURNS_ADDRESS;
!   }
    if (writebuf)
      iq2000_store_return_value (type, regcache, writebuf);
    else if (readbuf)
