*** gdb-10.2.aarch64-elf/gdb/m68hc11-tdep.c	2021-04-25 13:06:26.000000000 +0900
--- gdb-10.2.m68hc11-elf/gdb/m68hc11-tdep.c	2021-07-29 06:15:02.000000000 +0900
***************
*** 1293,1300 ****
    if (valtype->code () == TYPE_CODE_STRUCT
        || valtype->code () == TYPE_CODE_UNION
        || valtype->code () == TYPE_CODE_ARRAY
!       || TYPE_LENGTH (valtype) > 4)
!     return RETURN_VALUE_STRUCT_CONVENTION;
    else
      {
        if (readbuf != NULL)
--- 1293,1306 ----
    if (valtype->code () == TYPE_CODE_STRUCT
        || valtype->code () == TYPE_CODE_UNION
        || valtype->code () == TYPE_CODE_ARRAY
!       || TYPE_LENGTH (valtype) > 4) {
!     if (readbuf) {
!       ULONGEST addr;
!       regcache_raw_read_unsigned (regcache, HARD_D_REGNUM, &addr);
!       read_memory (addr, readbuf, TYPE_LENGTH(valtype));
!     }
!     return RETURN_VALUE_ABI_RETURNS_ADDRESS;
!   }
    else
      {
        if (readbuf != NULL)
