*** gcc-12.1.0.org/gcc/config.gcc	2022-07-10 10:30:45.000000000 +0900
--- gcc-12.1.0.nios2-elf.mine/gcc/config.gcc	2022-09-06 11:30:39.000000000 +0900
***************
*** 478,484 ****
  	;;
  nios2-*-*)
  	cpu_type=nios2
- 	extra_options="${extra_options} g.opt"
  	;;
  nvptx-*-*)
  	cpu_type=nvptx
--- 478,483 ----
***************
*** 2819,2826 ****
                  tm_file="${tm_file} gnu-user.h linux.h glibc-stdint.h nios2/linux.h "
                  ;;
  	nios2-*-elf*)
! 		tm_file="${tm_file} newlib-stdint.h nios2/elf.h"
! 		extra_options="${extra_options} nios2/elf.opt"
  		;;
  	nios2-*-rtems*)
  		tm_file="${tm_file} newlib-stdint.h nios2/rtems.h rtems.h"
--- 2818,2824 ----
                  tm_file="${tm_file} gnu-user.h linux.h glibc-stdint.h nios2/linux.h "
                  ;;
  	nios2-*-elf*)
! 		tm_file="${tm_file} newlib-stdint.h"
  		;;
  	nios2-*-rtems*)
  		tm_file="${tm_file} newlib-stdint.h nios2/rtems.h rtems.h"
