*** gcc-10.2.0.aarch64-elf/libgcc/config/arm/ieee754-df.S	2020-07-23 15:35:18.000000000 +0900
--- gcc-10.2.0.arm-elf/libgcc/config/arm/ieee754-df.S	2021-11-28 16:31:06.000000000 +0900
***************
*** 619,635 ****
  
  #if defined(L_arm_muldf3) || defined(L_arm_muldivdf3)
  
  @ Define multiplication as weak in _arm_muldf3.o so that it can be overriden
  @ by the global definition in _arm_muldivdf3.o.  This allows a program only
  @ using multiplication to take the weak definition which does not contain the
  @ division code. Programs using only division or both division and
  @ multiplication will pull _arm_muldivdf3.o from which both the multiplication
  @ and division are taken thanks to the override.
! #ifdef L_arm_muldf3
! WEAK muldf3
! WEAK aeabi_dmul
! #endif
  
  ARM_FUNC_START muldf3
  ARM_FUNC_ALIAS aeabi_dmul muldf3
  	CFI_START_FUNCTION
--- 619,638 ----
  
  #if defined(L_arm_muldf3) || defined(L_arm_muldivdf3)
  
+ 	
  @ Define multiplication as weak in _arm_muldf3.o so that it can be overriden
  @ by the global definition in _arm_muldivdf3.o.  This allows a program only
  @ using multiplication to take the weak definition which does not contain the
  @ division code. Programs using only division or both division and
  @ multiplication will pull _arm_muldivdf3.o from which both the multiplication
  @ and division are taken thanks to the override.
! // comment-outed by khasegawa
! // #ifdef L_arm_muldf3
! // .weak muldf3
! // .weak aeabi_dmul
! // #endif
  
+ #ifdef L_arm_muldf3    // add khasegawa
  ARM_FUNC_START muldf3
  ARM_FUNC_ALIAS aeabi_dmul muldf3
  	CFI_START_FUNCTION
***************
*** 713,719 ****
  	adcs	xl, xl, #0
  	adc	xh, xh, r4, lsl #20
  	RETLDM	"r4, r5, r6"
! 
  	@ Multiplication by 0x1p*: let''s shortcut a lot of code.
  LSYM(Lml_1):
  	and	r6, r6, #0x80000000
--- 716,722 ----
  	adcs	xl, xl, #0
  	adc	xh, xh, r4, lsl #20
  	RETLDM	"r4, r5, r6"
! #endif // add by khasegawa
  	@ Multiplication by 0x1p*: let''s shortcut a lot of code.
  LSYM(Lml_1):
  	and	r6, r6, #0x80000000
***************
*** 877,885 ****
--- 880,890 ----
  	orr	xh, xh, #0x00f80000
  	RETLDM	"r4, r5, r6"
  
+ #ifdef L_arm_muldf3 // add by khasegawa
  	CFI_END_FUNCTION
  	FUNC_END aeabi_dmul
  	FUNC_END muldf3
+ #endif  // add by khasegawa
  
  #ifdef L_arm_muldivdf3
  
