*** gcc-10.2.0.org/gcc/config.gcc	2021-10-16 10:31:18.000000000 +0900
--- gcc-10.2.0.or1k-elf.mine/gcc/config.gcc	2022-02-10 16:49:10.000000000 +0900
***************
*** 2818,2825 ****
  		tm_file="${tm_file} or1k/linux.h"
  		;;
  	or1k*-*-elf*)
! 		tm_file="${tm_file} newlib-stdint.h or1k/elf.h"
! 		extra_options="${extra_options} or1k/elf.opt"
  		;;
  	or1k*-*-rtems*)
  		tm_file="${tm_file} newlib-stdint.h or1k/rtems.h rtems.h"
--- 2818,2824 ----
  		tm_file="${tm_file} or1k/linux.h"
  		;;
  	or1k*-*-elf*)
! 		tm_file="${tm_file} newlib-stdint.h"
  		;;
  	or1k*-*-rtems*)
  		tm_file="${tm_file} newlib-stdint.h or1k/rtems.h rtems.h"
