*** gcc-10.2.0.org/gcc/config.gcc	2021-10-16 10:31:18.000000000 +0900
--- gcc-10.2.0.mmix-elf.mine/gcc/config.gcc	2022-04-02 17:08:06.000000000 +0900
***************
*** 1673,1678 ****
--- 1673,1682 ----
  	tm_file="${tm_file} dbxelf.h elfos.h moxie/moxiebox.h newlib-stdint.h"
  	tmake_file="${tmake_file} moxie/t-moxiebox"
  	;;
+ mmix-*-elf*)
+  	tmake_file="mmix/t-mmix"
+  	tm_file="dbxelf.h elfos.h mmix/mmix.h newlib-stdint.h"
+  	;;
  h8300-*-elf*)
  	tmake_file="h8300/t-h8300"
  	tm_file="h8300/h8300.h dbxelf.h elfos.h newlib-stdint.h h8300/elf.h"
