*** gcc-10.2.0.org/gcc/config.gcc	2021-10-16 10:31:18.000000000 +0900
--- gcc-10.2.0.m68k-elf.mine/gcc/config.gcc	2022-05-27 07:15:42.000000000 +0900
***************
*** 463,470 ****
  	extra_options="${extra_options} g.opt"
          ;;
  m68k-*-*)
- 	extra_headers=math-68881.h
- 	extra_options="${extra_options} m68k/m68k-tables.opt"
  	;;
  microblaze*-*-*)
          cpu_type=microblaze
--- 463,468 ----
***************
*** 2301,2335 ****
  	fi
  	;;
  m68k-*-elf* | fido-*-elf*)
! 	case ${target} in
! 	fido-*-elf*)
! 		# Check that $with_cpu makes sense.
! 		case $with_cpu in
! 		"" | "fidoa")
! 			;;
! 		*)
! 			echo "Cannot accept --with-cpu=$with_cpu"
! 			exit 1
! 			;;
! 		esac
! 		with_cpu=fidoa
! 		;;
! 	*)
! 		default_m68k_cpu=68020
! 		default_cf_cpu=5206
! 		;;
! 	esac
! 	tm_file="${tm_file} m68k/m68k-none.h m68k/m68kelf.h dbxelf.h elfos.h newlib-stdint.h m68k/m68kemb.h m68k/m68020-elf.h"
! 	tm_defines="${tm_defines} MOTOROLA=1"
! 	tmake_file="m68k/t-floatlib m68k/t-m68kbare m68k/t-m68kelf"
! 	# Add multilibs for targets other than fido.
! 	case ${target} in
! 	fido-*-elf*)
! 		;;
! 	*)
! 		tmake_file="$tmake_file m68k/t-mlibs"
! 		;;
! 	esac
  	;;
  m68k*-*-netbsdelf*)
  	default_m68k_cpu=68020
--- 2299,2307 ----
  	fi
  	;;
  m68k-*-elf* | fido-*-elf*)
! 	tm_file="${tm_file} dbxelf.h elfos.h newlib-stdint.h"
! 	tm_defines="${tm_defines}"
! 	tmake_file="m68k/t-m68k"
  	;;
  m68k*-*-netbsdelf*)
  	default_m68k_cpu=68020
***************
*** 4384,4398 ****
  			tm_defines="$tm_defines M68K_DEFAULT_TUNE=u68020_60"
  			;;
  		*)
- 			# We need the C identifier rather than the string.
- 			m68k_cpu_ident=`awk -v arg="\"$with_cpu\"" \
- 			   'BEGIN { FS="[ \t]*[,()][ \t]*" }; \
- 			    $1 == "M68K_DEVICE" && $2 == arg { print $3 }' \
- 				 ${srcdir}/config/m68k/m68k-devices.def`
- 			if [ x"$m68k_cpu_ident" = x ] ; then
- 				echo "Unknown CPU used in --with-cpu=$with_cpu" 1>&2
- 				exit 1
- 			fi
  			with_cpu="mcpu=$with_cpu"
  			;;
  		esac
--- 4356,4361 ----
