*** gdb-10.2.arm-elf/sim/aarch64/simulator.c	2021-04-25 13:06:26.000000000 +0900
--- gdb-10.2.aarch64-elf/sim/aarch64/simulator.c	2021-11-28 08:19:35.000000000 +0900
***************
*** 9411,9416 ****
--- 9419,9424 ----
  	case 0x01: do_scalar_MOV (cpu); return;
  	case 0x39: do_scalar_FCM (cpu); return;
  	case 0x3B: do_scalar_FCM (cpu); return;
+ 	case 0x36: scvtf(cpu); return;
  	}
        break;
