*** gcc-10.2.0.org/gcc/config.gcc	2021-10-16 10:31:18.000000000 +0900
--- gcc-10.2.0.i386-elf.mine/gcc/config.gcc	2022-04-21 11:05:29.000000000 +0900
***************
*** 382,419 ****
  	extra_headers=math-68881.h
  	extra_options="${extra_options} m68k/m68k-tables.opt"
          ;;
! i[34567]86-*-*)
  	cpu_type=i386
- 	c_target_objs="i386-c.o"
- 	cxx_target_objs="i386-c.o"
- 	d_target_objs="i386-d.o"
- 	extra_objs="x86-tune-sched.o x86-tune-sched-bd.o x86-tune-sched-atom.o x86-tune-sched-core.o i386-options.o i386-builtins.o i386-expand.o i386-features.o"
- 	target_gtfiles="\$(srcdir)/config/i386/i386-builtins.c \$(srcdir)/config/i386/i386-expand.c \$(srcdir)/config/i386/i386-options.c"
- 	extra_options="${extra_options} fused-madd.opt"
- 	extra_headers="cpuid.h mmintrin.h mm3dnow.h xmmintrin.h emmintrin.h
- 		       pmmintrin.h tmmintrin.h ammintrin.h smmintrin.h
- 		       nmmintrin.h bmmintrin.h fma4intrin.h wmmintrin.h
- 		       immintrin.h x86intrin.h avxintrin.h xopintrin.h
- 		       ia32intrin.h cross-stdarg.h lwpintrin.h popcntintrin.h
- 		       lzcntintrin.h bmiintrin.h bmi2intrin.h tbmintrin.h
- 		       avx2intrin.h avx512fintrin.h fmaintrin.h f16cintrin.h
- 		       rtmintrin.h xtestintrin.h rdseedintrin.h prfchwintrin.h
- 		       adxintrin.h fxsrintrin.h xsaveintrin.h xsaveoptintrin.h
- 		       avx512cdintrin.h avx512erintrin.h avx512pfintrin.h
- 		       shaintrin.h clflushoptintrin.h xsavecintrin.h
- 		       xsavesintrin.h avx512dqintrin.h avx512bwintrin.h
- 		       avx512vlintrin.h avx512vlbwintrin.h avx512vldqintrin.h
- 		       avx512ifmaintrin.h avx512ifmavlintrin.h avx512vbmiintrin.h
- 		       avx512vbmivlintrin.h avx5124fmapsintrin.h avx5124vnniwintrin.h
- 		       avx512vpopcntdqintrin.h clwbintrin.h mwaitxintrin.h
- 		       clzerointrin.h pkuintrin.h sgxintrin.h cetintrin.h
- 		       gfniintrin.h cet.h avx512vbmi2intrin.h
- 		       avx512vbmi2vlintrin.h avx512vnniintrin.h
- 		       avx512vnnivlintrin.h vaesintrin.h vpclmulqdqintrin.h
- 		       avx512vpopcntdqvlintrin.h avx512bitalgintrin.h
- 		       pconfigintrin.h wbnoinvdintrin.h movdirintrin.h
- 		       waitpkgintrin.h cldemoteintrin.h avx512bf16vlintrin.h avx512bf16intrin.h
- 		       enqcmdintrin.h avx512vp2intersectintrin.h avx512vp2intersectvlintrin.h"
  	;;
  x86_64-*-*)
  	cpu_type=i386
--- 382,389 ----
  	extra_headers=math-68881.h
  	extra_options="${extra_options} m68k/m68k-tables.opt"
          ;;
! i386-*-*)
  	cpu_type=i386
  	;;
  x86_64-*-*)
  	cpu_type=i386
***************
*** 1861,1868 ****
  i[34567]86-*-elfiamcu)
  	tm_file="${tm_file} i386/unix.h i386/att.h dbxelf.h elfos.h newlib-stdint.h i386/iamcu.h"
  	;;
! i[34567]86-*-elf*)
! 	tm_file="${tm_file} i386/unix.h i386/att.h dbxelf.h elfos.h newlib-stdint.h i386/i386elf.h"
  	;;
  x86_64-*-elf*)
  	tm_file="${tm_file} i386/unix.h i386/att.h dbxelf.h elfos.h newlib-stdint.h i386/i386elf.h i386/x86-64.h"
--- 1831,1842 ----
  i[34567]86-*-elfiamcu)
  	tm_file="${tm_file} i386/unix.h i386/att.h dbxelf.h elfos.h newlib-stdint.h i386/iamcu.h"
  	;;
! i386-*-elf*)
! 	tm_file="${tm_file} dbxelf.h elfos.h newlib-stdint.h"
!   	tm_p_file="i386/i386-protos.h"
!   	md_file="i386/i386.md"
!   	out_file="i386/i386.c"
!   	tmake_file="i386/t-i386"
  	;;
  x86_64-*-elf*)
  	tm_file="${tm_file} i386/unix.h i386/att.h dbxelf.h elfos.h newlib-stdint.h i386/i386elf.h i386/x86-64.h"
