*** gcc-12.1.0.org/gcc/config.gcc	2022-07-10 10:30:45.000000000 +0900
--- gcc-12.1.0.ia64-elf.mine/gcc/config.gcc	2022-07-28 08:41:17.000000000 +0900
***************
*** 426,433 ****
  		       mwaitintrin.h avx512fp16intrin.h avx512fp16vlintrin.h"
  	;;
  ia64-*-*)
- 	extra_headers=ia64intrin.h
- 	extra_options="${extra_options} g.opt fused-madd.opt"
  	;;
  hppa*-*-*)
  	cpu_type=pa
--- 426,431 ----
***************
*** 2247,2263 ****
  	tm_file="${tm_file} i386/unix.h i386/att.h elfos.h newlib-stdint.h i386/i386elf.h i386/x86-64.h fuchsia.h"
  	;;
  ia64*-*-elf*)
! 	tm_file="${tm_file} dbxelf.h elfos.h newlib-stdint.h ia64/sysv4.h ia64/elf.h"
  	tmake_file="ia64/t-ia64"
- 	target_cpu_default="0"
- 	if test x$gas = xyes
- 	then
- 		target_cpu_default="${target_cpu_default}|MASK_GNU_AS"
- 	fi
- 	if test x$gnu_ld = xyes
- 	then
- 		target_cpu_default="${target_cpu_default}|MASK_GNU_LD"
- 	fi
  	;;
  ia64*-*-freebsd*)
  	tm_file="${tm_file} dbxelf.h elfos.h ${fbsd_tm_file} ia64/sysv4.h ia64/freebsd.h"
--- 2245,2252 ----
  	tm_file="${tm_file} i386/unix.h i386/att.h elfos.h newlib-stdint.h i386/i386elf.h i386/x86-64.h fuchsia.h"
  	;;
  ia64*-*-elf*)
! 	tm_file="${tm_file} dbxelf.h elfos.h newlib-stdint.h"
  	tmake_file="ia64/t-ia64"
  	;;
  ia64*-*-freebsd*)
  	tm_file="${tm_file} dbxelf.h elfos.h ${fbsd_tm_file} ia64/sysv4.h ia64/freebsd.h"
