*** gcc-10.2.0.org/gcc/config.gcc	2021-10-16 10:31:18.000000000 +0900
--- gcc-10.2.0.h8300-elf.mine/gcc/config.gcc	2022-02-05 17:05:59.000000000 +0900
***************
*** 1675,1681 ****
  	;;
  h8300-*-elf*)
  	tmake_file="h8300/t-h8300"
! 	tm_file="h8300/h8300.h dbxelf.h elfos.h newlib-stdint.h h8300/elf.h"
  	;;
  h8300-*-linux*)
  	tmake_file="${tmake_file} h8300/t-h8300 h8300/t-linux"
--- 1675,1681 ----
  	;;
  h8300-*-elf*)
  	tmake_file="h8300/t-h8300"
! 	tm_file="h8300/h8300.h dbxelf.h elfos.h newlib-stdint.h"
  	;;
  h8300-*-linux*)
  	tmake_file="${tmake_file} h8300/t-h8300 h8300/t-linux"
