*** gcc-10.2.0.org/gcc/config.gcc	2021-10-16 10:31:18.000000000 +0900
--- gcc-10.2.0.pru-elf.mine/gcc/config.gcc	2022-02-10 17:14:42.000000000 +0900
***************
*** 3038,3044 ****
  pru*-*-*)
  	tm_file="elfos.h newlib-stdint.h ${tm_file}"
  	tmake_file="${tmake_file} pru/t-pru"
- 	extra_objs="pru-pragma.o pru-passes.o"
  	use_gcc_stdint=wrap
  	;;
  rs6000-ibm-aix6.* | powerpc-ibm-aix6.*)
--- 3038,3043 ----
