*** gdb-10.2.org/gdb/infrun.c	2021-04-25 13:06:26.000000000 +0900
--- gdb-10.2.mn10300-elf/gdb/infrun.c	2021-12-25 12:08:55.000000000 +0900
***************
*** 5782,5787 ****
--- 5782,5795 ----
  
  /* Come here when the program has stopped with a signal.  */
  
+ #undef target_stopped_by_sw_breakpoint
+ bool target_stopped_by_sw_breakpoint(CORE_ADDR pc)
+ {
+   regcache* regcache = get_current_regcache();
+   const address_space* aspace = regcache->aspace();
+   return breakpoint_here_p(aspace, pc) == no_breakpoint_here;
+ }
+ 
  static void
  handle_signal_stop (struct execution_control_state *ecs)
  {
***************
*** 6116,6122 ****
  
    /* Maybe this was a trap for a software breakpoint that has since
       been removed.  */
!   if (random_signal && target_stopped_by_sw_breakpoint ())
      {
        if (gdbarch_program_breakpoint_here_p (gdbarch,
  					     ecs->event_thread->suspend.stop_pc))
--- 6124,6130 ----
  
    /* Maybe this was a trap for a software breakpoint that has since
       been removed.  */
!   if (random_signal && target_stopped_by_sw_breakpoint (ecs->event_thread->suspend.stop_pc))
      {
        if (gdbarch_program_breakpoint_here_p (gdbarch,
  					     ecs->event_thread->suspend.stop_pc))
*** gdb-10.2.org/gdb/mn10300-tdep.c	2021-04-25 13:06:26.000000000 +0900
--- gdb-10.2.mn10300-elf/gdb/mn10300-tdep.c	2021-12-25 12:05:20.000000000 +0900
***************
*** 238,245 ****
  		      struct type *type, struct regcache *regcache,
  		      gdb_byte *readbuf, const gdb_byte *writebuf)
  {
!   if (mn10300_use_struct_convention (type))
!     return RETURN_VALUE_STRUCT_CONVENTION;
  
    if (readbuf)
      mn10300_extract_return_value (gdbarch, type, regcache, readbuf);
--- 238,251 ----
  		      struct type *type, struct regcache *regcache,
  		      gdb_byte *readbuf, const gdb_byte *writebuf)
  {
!   if (mn10300_use_struct_convention (type)) {
!     if (readbuf) {
!       ULONGEST addr;
!       regcache_raw_read_unsigned(regcache, 4, &addr);
!       read_memory(addr, readbuf, TYPE_LENGTH(type));
!     }
!     return RETURN_VALUE_ABI_RETURNS_ADDRESS;
!   }
  
    if (readbuf)
      mn10300_extract_return_value (gdbarch, type, regcache, readbuf);
***************
*** 1404,1409 ****
--- 1410,1417 ----
    /* Hook in ABI-specific overrides, if they have been registered.  */
    gdbarch_init_osabi (info, gdbarch);
  
+   set_gdbarch_decr_pc_after_break(gdbarch, sizeof mn10300_break_insn);
+     
    return gdbarch;
  }
   
