*** gcc-10.2.0.org/gcc/config.gcc	2021-10-16 10:31:18.000000000 +0900
--- gcc-10.2.0.msp430-elf.mine/gcc/config.gcc	2022-02-09 09:20:21.000000000 +0900
***************
*** 2701,2711 ****
  	;;
  msp430-*-*)
  	tm_file="dbxelf.h elfos.h newlib-stdint.h ${tm_file}"
- 	c_target_objs="msp430-c.o"
- 	cxx_target_objs="msp430-c.o"
  	tmake_file="${tmake_file} msp430/t-msp430"
- 	extra_objs="${extra_objs} msp430-devices.o"
- 	extra_gcc_objs="driver-msp430.o msp430-devices.o"
  	# Enable .init_array unless it has been explicitly disabled.
  	# The MSP430 EABI mandates the use of .init_array, and the Newlib CRT
  	# code since mid-2019 expects it.
--- 2701,2707 ----
