*** gcc-12.1.0.org/gcc/config.gcc	2022-07-10 10:30:45.000000000 +0900
--- gcc-12.1.0.epiphany-elf.mine/gcc/config.gcc	2022-07-12 14:23:56.000000000 +0900
***************
*** 1638,1647 ****
  	  tm_file="${tm_file} newlib-stdint.h"
  	  ;;
  	esac
- 	extra_options="${extra_options} fused-madd.opt"
- 	extra_objs="${extra_objs} mode-switch-use.o resolve-sw-modes.o"
- 	tm_defines="${tm_defines} EPIPHANY_STACK_OFFSET=${with_stack_offset:-8}"
- 	extra_headers="epiphany_intrinsics.h"
  	;;
  fr30-*-elf)
  	tm_file="dbxelf.h elfos.h newlib-stdint.h ${tm_file}"
--- 1638,1643 ----
