*** gdb-10.2.h8300-elf/opcodes/mcore-dis.c	2021-04-25 13:06:26.000000000 +0900
--- gdb-10.2.mcore-elf/opcodes/mcore-dis.c	2021-07-26 08:42:50.000000000 +0900
***************
*** 28,34 ****
  #include "disassemble.h"
  
  /* Mask for each mcore_opclass: */
! static const unsigned short imsk[] = {
      /* O0  */ 0xFFFF,
      /* OT  */ 0xFFFC,
      /* O1  */ 0xFFF0,
--- 28,35 ----
  #include "disassemble.h"
  
  /* Mask for each mcore_opclass: */
! //static
! const unsigned short imsk[] = {
      /* O0  */ 0xFFFF,
      /* OT  */ 0xFFFC,
      /* O1  */ 0xFFF0,
