*** gcc-10.2.0.org/gcc/config.gcc	2021-10-16 10:31:18.000000000 +0900
--- gcc-10.2.0.microblaze-elf.mine/gcc/config.gcc	2022-02-08 10:18:26.000000000 +0900
***************
*** 2423,2431 ****
  			tm_defines="${tm_defines} TARGET_BIG_ENDIAN_DEFAULT=4321"
  			;;
  	esac
! 	tm_file="${tm_file} dbxelf.h newlib-stdint.h"
! 	c_target_objs="${c_target_objs} microblaze-c.o"
! 	cxx_target_objs="${cxx_target_objs} microblaze-c.o"
  	tmake_file="${tmake_file} microblaze/t-microblaze"
          ;;
  riscv*-*-linux*)
--- 2423,2429 ----
  			tm_defines="${tm_defines} TARGET_BIG_ENDIAN_DEFAULT=4321"
  			;;
  	esac
! 	tm_file="${tm_file} dbxelf.h elfos.h newlib-stdint.h"
  	tmake_file="${tmake_file} microblaze/t-microblaze"
          ;;
  riscv*-*-linux*)
