*** gdb-10.2.org/gdb/or1k-tdep.c	2021-04-25 13:06:26.000000000 +0900
--- gdb-10.2.or1k-elf/gdb/or1k-tdep.c	2021-12-29 09:36:12.000000000 +0900
***************
*** 567,572 ****
--- 567,573 ----
  }
  
  /* Implement the push_dummy_code gdbarch method.  */
+ #include "objfiles.h"
  
  static CORE_ADDR
  or1k_push_dummy_code (struct gdbarch *gdbarch, CORE_ADDR sp,
***************
*** 574,579 ****
--- 575,581 ----
  		      struct type *value_type, CORE_ADDR * real_pc,
  		      CORE_ADDR * bp_addr, struct regcache *regcache)
  {
+ #if 0  
    CORE_ADDR bp_slot;
  
    /* Reserve enough room on the stack for our breakpoint instruction.  */
***************
*** 586,591 ****
--- 588,598 ----
    *real_pc = function;
  
    return sp;
+ #else
+   *bp_addr = entry_point_address() + 8;
+   *real_pc = function;
+   return sp;
+ #endif
  }
  
  /* Implement the push_dummy_call gdbarch method.  */
***************
*** 610,616 ****
--- 617,627 ----
    struct type *func_type = value_type (function);
  
    /* Return address */
+ #if 0 
    regcache_cooked_write_unsigned (regcache, OR1K_LR_REGNUM, bp_addr);
+ #else
+   regcache_cooked_write_unsigned (regcache, OR1K_LR_REGNUM, bp_addr - 8);
+ #endif  
  
    /* Register for the next argument.  */
    argreg = OR1K_FIRST_ARG_REGNUM;
***************
*** 1166,1173 ****
    set_gdbarch_skip_prologue (gdbarch, or1k_skip_prologue);
    set_gdbarch_inner_than (gdbarch, core_addr_lessthan);
    set_gdbarch_frame_align (gdbarch, or1k_frame_align);
    set_gdbarch_frame_red_zone_size (gdbarch, OR1K_FRAME_RED_ZONE_SIZE);
! 
    /* Functions to access frame data */
    set_gdbarch_unwind_pc (gdbarch, or1k_unwind_pc);
    set_gdbarch_unwind_sp (gdbarch, or1k_unwind_sp);
--- 1177,1186 ----
    set_gdbarch_skip_prologue (gdbarch, or1k_skip_prologue);
    set_gdbarch_inner_than (gdbarch, core_addr_lessthan);
    set_gdbarch_frame_align (gdbarch, or1k_frame_align);
+ #if 0  
    set_gdbarch_frame_red_zone_size (gdbarch, OR1K_FRAME_RED_ZONE_SIZE);
! #endif
!   
    /* Functions to access frame data */
    set_gdbarch_unwind_pc (gdbarch, or1k_unwind_pc);
    set_gdbarch_unwind_sp (gdbarch, or1k_unwind_sp);
