*** binutils-2.36.1.aarch64-elf/bfd/elf32-msp430.c	2021-01-09 19:47:33.000000000 +0900
--- binutils-2.36.1.msp430-elf/bfd/elf32-msp430.c	2022-02-13 11:00:24.000000000 +0900
***************
*** 2488,2494 ****
  		*again = TRUE;
  	      }
  	  }
! 
  	/* Try to turn a 16-bit absolute branch into a 10-bit pc-relative
  	   branch.  */
  	if ((uses_msp430x_relocs (abfd)
--- 2488,2494 ----
  		*again = TRUE;
  	      }
  	  }
! #if 0
  	/* Try to turn a 16-bit absolute branch into a 10-bit pc-relative
  	   branch.  */
  	if ((uses_msp430x_relocs (abfd)
***************
*** 2601,2606 ****
--- 2601,2607 ----
  		*again = TRUE;
  	      }
  	  }
+ #endif	
        }
  
    if (isymbuf != NULL && symtab_hdr->contents != (unsigned char *) isymbuf)
