*** gcc-10.2.0.org/gcc/config.gcc	2021-10-16 10:31:18.000000000 +0900
--- gcc-10.2.0.m32c-elf.mine/gcc/config.gcc	2022-02-07 08:13:07.000000000 +0900
***************
*** 3502,3509 ****
   	;;
  m32c-*-elf*)
  	tm_file="dbxelf.h elfos.h newlib-stdint.h ${tm_file}"
- 	c_target_objs="m32c-pragma.o"
- 	cxx_target_objs="m32c-pragma.o"
   	;;
  *)
  	echo "*** Configuration ${target} not supported" 1>&2
--- 3502,3507 ----
