*** gcc-10.2.0.org/gcc/config.gcc	2021-10-16 10:31:18.000000000 +0900
--- gcc-10.2.0.rx-elf.mine/gcc/config.gcc	2022-02-15 14:08:49.000000000 +0900
***************
*** 3078,3084 ****
  rx-*-elf*)
  	tm_file="dbxelf.h elfos.h newlib-stdint.h ${tm_file}"
  	tmake_file="${tmake_file} rx/t-rx"
- 	extra_options="${extra_options} rx/elf.opt"
  	;;
  rx-*-linux*)
  	tm_file="elfos.h linux.h glibc-stdint.h rx/linux.h ../../libgcc/config/rx/rx-abi.h"
--- 3078,3083 ----
