*** gcc-10.2.0.org/gcc/config.gcc	2021-10-16 10:31:18.000000000 +0900
--- gcc-10.2.0.mcore-elf.mine/gcc/config.gcc	2022-02-07 16:45:40.000000000 +0900
***************
*** 2378,2386 ****
  	tm_defines="${tm_defines} MOTOROLA=1"
  	;;
  mcore-*-elf)
! 	tm_file="dbxelf.h elfos.h newlib-stdint.h ${tm_file} mcore/mcore-elf.h"
  	tmake_file=mcore/t-mcore
- 	inhibit_libc=true
  	;;
  microblaze*-linux*)
  	case $target in
--- 2378,2385 ----
  	tm_defines="${tm_defines} MOTOROLA=1"
  	;;
  mcore-*-elf)
! 	tm_file="dbxelf.h elfos.h newlib-stdint.h ${tm_file}"
  	tmake_file=mcore/t-mcore
  	;;
  microblaze*-linux*)
  	case $target in
